From 200a1e6a73a80a29631d5e35a8b333672b0e8b7a Mon Sep 17 00:00:00 2001
From: Xueqiang Gong <quic_gongx@quicinc.com>
Date: Fri, 10 Nov 2023 09:29:30 +0800
Subject: [PATCH] Kernel: Add pcie and device tree changes to support hst/hsp chip

Add small configuations to support hst/hsp on third platform kernel:
1. Add device tree configuration for platform driver probe
2. Increase maximum supported MSI number to 256
3. Export function API

Signed-off-by: Xueqiang Gong <quic_gongx@quicinc.com>
---
 arch/arm64/boot/dts/freescale/fsl-imx8dx.dtsi | 42 +++++++++++++++++++
 arch/arm64/boot/dts/freescale/fsl-imx8mq.dtsi | 42 +++++++++++++++++++
 drivers/pci/dwc/pcie-designware.h             |  2 +-
 fs/read_write.c                               |  1 +
 4 files changed, 86 insertions(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8dx.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8dx.dtsi
index 72040461e67a..7cad3e641bbb 100644
--- a/arch/arm64/boot/dts/freescale/fsl-imx8dx.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8dx.dtsi
@@ -3412,6 +3412,11 @@
 		compatible = "fsl,imx8qxp-ocotp", "syscon";
 	};
 
+	mhi: qcom,mhi {
+		compatible = "qcom,mhi";
+		status = "okay";
+	};
+	
 	pcieb: pcie@0x5f010000 {
 		/*
 		 * pcieb phyx1 lane1 in default, adjust it refer to the
@@ -3462,6 +3467,43 @@
 		status = "disabled";
 	};
 
+	ipc_router: qcom,ipc_router_external_modem_xprt2 {
+		compatible = "qcom,ipc_router_mhi_xprt";
+		qcom,ch-name = "IPCRTR";
+		qcom,out-chan-id = <20>;
+		qcom,in-chan-id = <21>;
+		qcom,xprt-remote = "external-wlan";
+		qcom,xprt-linkid = <1>;
+		qcom,xprt-version = <3>;
+		status = "okay";
+	};
+
+	diag_mhi: qcom,diag-mhi {
+		compatible = "qcom,diag-mhi";
+		status = "okay";
+	};
+
+       cnss2: qcom,cnss-qca-converged {
+               compatible = "qcom,cnss-qca-converged";
+       
+               pinctrl-names = "default";
+               pinctrl-0 = <&pinctrl_wifi>;
+
+               qcom,converged-dt;
+               qcom,bus-type=<0>;
+               status = "okay";
+
+               /* Support Hasting Device */
+               chip_cfg@1 {
+                       supported-ids = <0x1101>;
+               };
+
+               /* Support Hasting Prime Device */
+               chip_cfg@3 {
+                       supported-ids = <0x1103>;
+               };
+       };
+	
 	imx_ion {
 		compatible = "fsl,mxc-ion";
 		fsl,heap-id = <0>;
diff --git a/arch/arm64/boot/dts/freescale/fsl-imx8mq.dtsi b/arch/arm64/boot/dts/freescale/fsl-imx8mq.dtsi
index a1d100401418..892f2d219175
--- a/arch/arm64/boot/dts/freescale/fsl-imx8mq.dtsi
+++ b/arch/arm64/boot/dts/freescale/fsl-imx8mq.dtsi
@@ -1295,6 +1295,11 @@
 		status = "disabled";
 	};
 
+	mhi: qcom,mhi {
+		compatible = "qcom,mhi";
+		status = "okay";
+	};
+
 	pcie0: pcie@0x33800000 {
 		compatible = "fsl,imx8mq-pcie", "snps,dw-pcie";
 		reg = <0x0 0x33800000 0x0 0x400000>, <0x0 0x1ff00000 0x0 0x80000>;
@@ -1355,6 +1360,43 @@
 		status = "disabled";
 	};
 
+	ipc_router: qcom,ipc_router_external_modem_xprt2 {
+		compatible = "qcom,ipc_router_mhi_xprt";
+		qcom,ch-name = "IPCRTR";
+		qcom,out-chan-id = <20>;
+		qcom,in-chan-id = <21>;
+		qcom,xprt-remote = "external-wlan";
+		qcom,xprt-linkid = <1>;
+		qcom,xprt-version = <3>;
+		status = "okay";
+	};
+
+	diag_mhi: qcom,diag-mhi {
+		compatible = "qcom,diag-mhi";
+		status = "okay";
+	};
+
+       cnss2: qcom,cnss-qca-converged {
+               compatible = "qcom,cnss-qca-converged";
+       
+               pinctrl-names = "default";
+
+
+               qcom,converged-dt;
+               qcom,bus-type=<0>;
+               status = "okay";
+
+               /* Support Hasting Device */
+               chip_cfg@1 {
+                       supported-ids = <0x1101>;
+               };
+
+               /* Support Hasting Prime Device */
+               chip_cfg@3 {
+                       supported-ids = <0x1103>;
+               };
+       };
+
 	ddr_pmu0: ddr_pmu@3d800000 {
 		compatible = "fsl,imx8-ddr-pmu";
 		reg = <0x0 0x3d800000 0x0 0x400000>;
diff --git a/drivers/pci/dwc/pcie-designware.h b/drivers/pci/dwc/pcie-designware.h
index ef0bd7fb8089..e9f67f2a016a 100644
--- a/drivers/pci/dwc/pcie-designware.h
+++ b/drivers/pci/dwc/pcie-designware.h
@@ -110,7 +110,7 @@
  * it 32 as of now. Probably we will never need more than 32. If needed,
  * then increment it in multiple of 32.
  */
-#define MAX_MSI_IRQS			32
+#define MAX_MSI_IRQS			256
 #define MAX_MSI_CTRLS			(MAX_MSI_IRQS / 32)
 
 struct pcie_port;
diff --git a/fs/read_write.c b/fs/read_write.c
index 57a00ef895b2..604ce08c0eaf 100644
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -552,6 +552,7 @@ ssize_t vfs_write(struct file *file, const char __user *buf, size_t count, loff_
 
 	return ret;
 }
+EXPORT_SYMBOL(vfs_write);
 
 static inline loff_t file_pos_read(struct file *file)
 {
-- 
2.17.1

